cpp/src/arrow/python/inference.cc

```diff
@@ -354,12 +354,17 @@ class TypeInferrer {
       *keep_going = make_unions_;
     } else if (PyArray_CheckAnyScalarExact(obj)) {
       RETURN_NOT_OK(VisitDType(PyArray_DescrFromScalar(obj), keep_going));
-    } else if (PyList_Check(obj)) {
-      RETURN_NOT_OK(VisitList(obj, keep_going));
+    } else if (PySet_Check(obj) || (Py_TYPE(obj) == &PyDictValues_Type)) {
+      RETURN_NOT_OK(VisitSet(obj, keep_going));
     } else if (PyArray_Check(obj)) {
       RETURN_NOT_OK(VisitNdarray(obj, keep_going));
     } else if (PyDict_Check(obj)) {
       RETURN_NOT_OK(VisitDict(obj));
+    } else if (PySequence_Check(obj)) {
+      // Deals with generic sequences. Including list and tuple.
```

Comment:
I'm not fond of allowing generic sequences here, it might allow all kinds of types that we don't want to convert implicitly. Explicitly checking list and tuple would be better.

Reply:
The `PySequence_Check` is necessary to ensure we support `collections.deque` which was supported at top level (`array(deque)`) but not at nested levels `array([deque])`. Using `PySequence` here adds support for deques too.

I switched to checking explicitly the `deque` type instead of using `PySequence_Check`, so now the check should be more explicit.

Reply:
Hmm. I think trying to support common collections types is a bit of a slippery slope. We could also limit ourselves to builtins (list, set, tuple).

Reply:
Now it does check for explicitly supported types (`tuple`, `list`, `deque`) instead of `PySequence`